stdlib/list.ml

```diff
@@ -125,10 +157,26 @@ let rec fold_right f l accu =
     [] -> accu
   | a::l -> f a (fold_right f l accu)
 
-let rec map2 f l1 l2 =
+let[@tail_mod_cons] rec map2 f l1 l2 =
   match (l1, l2) with
     ([], []) -> []
-  | (a1::l1, a2::l2) -> let r = f a1 a2 in r :: map2 f l1 l2
+  | ([a1], [b1]) ->
+      let r1 = f a1 b1 in
+      [r1]
+  | ([a1; a2], [b1; b2]) ->
+      let r1 = f a1 b1 in
+      let r2 = f a2 b2 in
+      [r1; r2]
+  | ([a1; a2; a3], [b1; b2; b3]) ->
+      let r1 = f a1 b1 in
+      let r2 = f a2 b2 in
+      let r3 = f a3 b3 in
+      [r1; r2; r3]
+  | (a1::a2::a3::l1, b1::b2::b3::l2) ->
+      let r1 = f a1 b1 in
+      let r2 = f a2 b2 in
+      let r3 = f a3 b3 in
+      r1::r2::r3::map2 f l1 l2
   | (_, _) -> invalid_arg "List.map2"
```

Comment:
I wonder what is the impact of unrolling on code size in this case of partial pattern matching followed by a catch-all. Risks I see include (1) sensibly more pattern-matching-related code being produced (due to unrolled cases not being an exhaustive covering) and (2) performance overhead. In your stead I would (1) ensure that you have a benchmark for this one compared to the non-unrolled variant, (2) eyeball the -dlambda or -dcmm output to ensure that there is no blowup, (3) consider not unrolling this one.

Reply:
Good point. Re performance numbers below (roughly unrolled TRMC > stdlib > TRMC, bigger is faster). For code size indeed there seems to be quite an increase. I wasn't sure how to quantify it, but the linear code is 107 lines for the TRMC (no unrolling) vs 420 lines for TRMC (with unrolling).

```
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌──────────────────────┬──────────┬─────────┬────────────┐
│ Name                 │ Time/Run │ mWd/Run │ Percentage │
├──────────────────────┼──────────┼─────────┼────────────┤
│ map2 trmc unrolled 1 │   5.78ns │   3.00w │     67.52% │
│ map2 trmc          1 │   8.56ns │   3.00w │    100.00% │
│ map2 stdlib        1 │   7.09ns │   3.00w │     82.92% │
└──────────────────────┴──────────┴─────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌──────────────────────┬──────────┬─────────┬────────────┐
│ Name                 │ Time/Run │ mWd/Run │ Percentage │
├──────────────────────┼──────────┼─────────┼────────────┤
│ map2 trmc unrolled 2 │   8.73ns │   6.00w │     72.10% │
│ map2 trmc          2 │  12.11ns │   6.00w │    100.00% │
│ map2 stdlib        2 │  10.32ns │   6.00w │     85.21% │
└──────────────────────┴──────────┴─────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌──────────────────────┬──────────┬─────────┬────────────┐
│ Name                 │ Time/Run │ mWd/Run │ Percentage │
├──────────────────────┼──────────┼─────────┼────────────┤
│ map2 trmc unrolled 3 │  10.80ns │   9.00w │     64.55% │
│ map2 trmc          3 │  16.73ns │   9.00w │    100.00% │
│ map2 stdlib        3 │  13.61ns │   9.00w │     81.33% │
└──────────────────────┴──────────┴─────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌──────────────────────┬──────────┬─────────┬────────────┐
│ Name                 │ Time/Run │ mWd/Run │ Percentage │
├──────────────────────┼──────────┼─────────┼────────────┤
│ map2 trmc unrolled 4 │  15.35ns │  12.00w │     74.18% │
│ map2 trmc          4 │  20.69ns │  12.00w │    100.00% │
│ map2 stdlib        4 │  17.10ns │  12.00w │     82.67% │
└──────────────────────┴──────────┴─────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌──────────────────────┬──────────┬─────────┬────────────┐
│ Name                 │ Time/Run │ mWd/Run │ Percentage │
├──────────────────────┼──────────┼─────────┼────────────┤
│ map2 trmc unrolled 5 │  17.18ns │  15.00w │     71.54% │
│ map2 trmc          5 │  24.01ns │  15.00w │    100.00% │
│ map2 stdlib        5 │  19.55ns │  15.00w │     81.42% │
└──────────────────────┴──────────┴─────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌──────────────────────┬──────────┬─────────┬────────────┐
│ Name                 │ Time/Run │ mWd/Run │ Percentage │
├──────────────────────┼──────────┼─────────┼────────────┤
│ map2 trmc unrolled 6 │  20.89ns │  18.00w │     74.09% │
│ map2 trmc          6 │  28.19ns │  18.00w │    100.00% │
│ map2 stdlib        6 │  23.11ns │  18.00w │     81.96% │
└──────────────────────┴──────────┴─────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌──────────────────────┬──────────┬─────────┬────────────┐
│ Name                 │ Time/Run │ mWd/Run │ Percentage │
├──────────────────────┼──────────┼─────────┼────────────┤
│ map2 trmc unrolled 7 │  23.94ns │  21.00w │     73.87% │
│ map2 trmc          7 │  32.41ns │  21.00w │    100.00% │
│ map2 stdlib        7 │  27.28ns │  21.00w │     84.17% │
└──────────────────────┴──────────┴─────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌──────────────────────┬──────────┬─────────┬────────────┐
│ Name                 │ Time/Run │ mWd/Run │ Percentage │
├──────────────────────┼──────────┼─────────┼────────────┤
│ map2 trmc unrolled 8 │  28.64ns │  24.00w │     78.83% │
│ map2 trmc          8 │  36.34ns │  24.00w │    100.00% │
│ map2 stdlib        8 │  31.61ns │  24.00w │     87.00% │
└──────────────────────┴──────────┴─────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌──────────────────────┬──────────┬─────────┬────────────┐
│ Name                 │ Time/Run │ mWd/Run │ Percentage │
├──────────────────────┼──────────┼─────────┼────────────┤
│ map2 trmc unrolled 9 │  31.31ns │  27.00w │     72.45% │
│ map2 trmc          9 │  43.22ns │  27.00w │    100.00% │
│ map2 stdlib        9 │  35.82ns │  27.00w │     82.88% │
└──────────────────────┴──────────┴─────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌───────────────────────┬──────────┬─────────┬────────────┐
│ Name                  │ Time/Run │ mWd/Run │ Percentage │
├───────────────────────┼──────────┼─────────┼────────────┤
│ map2 trmc unrolled 10 │  33.03ns │  30.00w │     74.05% │
│ map2 trmc          10 │  44.60ns │  30.00w │    100.00% │
│ map2 stdlib        10 │  36.78ns │  30.00w │     82.47% │
└───────────────────────┴──────────┴─────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌────────────────────────┬──────────┬─────────┬──────────┬──────────┬────────────┐
│ Name                   │ Time/Run │ mWd/Run │ mjWd/Run │ Prom/Run │ Percentage │
├────────────────────────┼──────────┼─────────┼──────────┼──────────┼────────────┤
│ map2 trmc unrolled 100 │ 327.20ns │ 300.00w │    0.67w │    0.67w │     74.92% │
│ map2 trmc          100 │ 436.73ns │ 300.00w │    0.68w │    0.68w │    100.00% │
│ map2 stdlib        100 │ 388.96ns │ 300.00w │    0.51w │    0.51w │     89.06% │
└────────────────────────┴──────────┴─────────┴──────────┴──────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌─────────────────────────┬──────────┬─────────┬──────────┬──────────┬────────────┐
│ Name                    │ Time/Run │ mWd/Run │ mjWd/Run │ Prom/Run │ Percentage │
├─────────────────────────┼──────────┼─────────┼──────────┼──────────┼────────────┤
│ map2 trmc unrolled 1000 │   3.84us │  3.00kw │   69.90w │   69.90w │     76.97% │
│ map2 trmc          1000 │   4.99us │  3.00kw │   69.90w │   69.90w │    100.00% │
│ map2 stdlib        1000 │   4.52us │  3.00kw │   52.22w │   52.22w │     90.57% │
└─────────────────────────┴──────────┴─────────┴──────────┴──────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌──────────────────────────┬──────────┬─────────┬──────────┬──────────┬────────────┐
│ Name                     │ Time/Run │ mWd/Run │ mjWd/Run │ Prom/Run │ Percentage │
├──────────────────────────┼──────────┼─────────┼──────────┼──────────┼────────────┤
│ map2 trmc unrolled 10000 │  96.55us │ 30.00kw │   7.54kw │   7.54kw │     93.08% │
│ map2 trmc          10000 │ 103.73us │ 30.00kw │   7.47kw │   7.47kw │    100.00% │
│ map2 stdlib        10000 │  98.35us │ 30.00kw │   5.57kw │   5.57kw │     94.81% │
└──────────────────────────┴──────────┴─────────┴──────────┴──────────┴────────────┘
Estimated testing time 30s (3 benchmarks x 10s). Change using '-quota'.
┌───────────────────────────┬──────────┬──────────┬──────────┬──────────┬────────────┐
│ Name                      │ Time/Run │  mWd/Run │ mjWd/Run │ Prom/Run │ Percentage │
├───────────────────────────┼──────────┼──────────┼──────────┼──────────┼────────────┤
│ map2 trmc unrolled 100000 │   2.76ms │ 300.00kw │ 299.97kw │ 299.97kw │     78.35% │
│ map2 trmc          100000 │   2.89ms │ 300.00kw │ 300.08kw │ 300.08kw │     82.08% │
│ map2 stdlib        100000 │   3.52ms │ 300.00kw │ 300.09kw │ 300.09kw │    100.00% │
└───────────────────────────┴──────────┴──────────┴──────────┴──────────┴────────────┘
```

Reply:
I think that's okay: the 3-unrolling is not much more than 3x the size of the no-unrolling version, and the performance profile is on par with the map1 versions. I think that your choice of using the same approach for `map2` is justified by these numbers.